### PR TITLE
IGNITE-21156 : [IEP-114] Custom metrics introduction

### PR DESCRIPTION
**Pull request by Vladsz83**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java [hunk 1]
```diff
@@ -101,32 +84,34 @@ public void registerMetrics(MetricRegistry discoReg) {
 
         discoReg.register("TotalReceivedMessages", this::totalReceivedMessages, "Total received messages count");
 
-        discoReg.register(joinedNodesCnt);
-        discoReg.register(failedNodesCnt);
-        discoReg.register(leftNodesCnt);
-        discoReg.register(pendingMsgsRegistered);
-        discoReg.register(rejectedSslConnectionsCnt);
+        discoReg.register(metricName(DISCO_METRICS, "JoinedNodes"), this::joinedNodesCount, "Joined nodes count");
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/spi/discovery/tcp/internal/TcpDiscoveryStatistics.java:87]: Can we just pass `MetricRegistryImpl` as a parameter and revert other changes?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistryImpl.java [hunk 2]
```diff
@@ -40,23 +41,21 @@
...
 import static org.apache.ignite.internal.processors.metric.impl.MetricUtils.fromFullName;
+import static org.apache.ignite.internal.processors.metric.impl.MetricUtils.isCustomPref;
 import static org.apache.ignite.internal.processors.metric.impl.MetricUtils.metricName;
 import static org.apache.ignite.internal.util.lang.GridFunc.nonThrowableSupplier;
 
 /**
  * Metric registry.
  */
-public class MetricRegistry implements ReadOnlyMetricRegistry {
+public class MetricRegistryImpl implements MetricRegistry {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistryImpl.java:58]: Looks like `MetricRegistry` only used when custom metrics returned to the user.
Looke like we can keep `MetricRegistryImpl implements ReadOnlyMetricRegistry` to reduce changes in internal code and return some kind of adapter to the user:

```
MetricRegistryImpl reg = ...;
return new MetricRegistry {
    public void register(...) {
        reg.register(...);
    }
}

File: modules/zookeeper/src/main/java/org/apache/ignite/spi/discovery/zk/internal/ZookeeperDiscoveryStatistics.java [hunk 3]
```diff
@@ -16,67 +16,59 @@
...
     /**
      * @param discoReg Discovery metric registry.
      */
     public void registerMetrics(MetricRegistry discoReg) {
-        discoReg.register(joinedNodesCnt);
-        discoReg.register(failedNodesCnt);
-        discoReg.register(leftNodesCnt);
-        discoReg.register(commErrCnt);
-        discoReg.register(topVer);
+        discoReg.register("JoinedNodes", this::joinedNodesCnt, "Joined nodes count");
```
Comment (nizhikov) [inline modules/zookeeper/src/main/java/org/apache/ignite/spi/discovery/zk/internal/ZookeeperDiscoveryStatistics.java:47]: Can we pass `MetricRegistryImpl` to reduce changes?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistryImpl.java (hunk 2 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/MetricRegistryImpl.java:58]: > Looks like MetricRegistry only used when custom metrics returned to the user.
It is used now in many places for the internal metrics. For instance, you can try searching for local variable declaration. 

Splitting implementations isn't what we wanted to do. We planned to implement custom metrics not with a single ticket. We deciced to add more methods to `MetricRegistry` like:
    `IntValueMetric intMetric(String name, @Nullable String desc);`
    `LongValueMetric longMetric(String name, @Nullable String desc);`
    `LongSumMetric longAdderMetric(String name, @Nullable String desc);`
    `DoubleValueMetric doubleMetric(String name, @Nullable String desc);`
in next ticket and keep the same metric regystry interface to cover also IGNITE-12553 (one of next tickets too) with `IgniteMetrics#findRegistry()`. 

If we split `class MetricRegistryImpl` and `interface MetricRegistry`, we will revert it if we want to expose read-only internal metrics by `IgniteMetrics#findRegistry()` in future.

If we want to add more metric registration methods to the interface, the adapter would grow with the next tickets.

File: modules/core/src/main/java/org/apache/ignite/metric/IgniteMetrics.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+import org.apache.ignite.lang.IgniteExperimental;
+import org.apache.ignite.spi.metric.ReadOnlyMetricRegistry;
+
+/**
+ * Allows to manage custom metrics.
+ * <p>
+ * Metrics are grouped into registries (groups). Every metric has full name which is the conjunction of registry name
+ * and the metric short name. Within a registry metric has only its own short name.
+ * <p>
+ * Note: Names of custom metric registries are required to start with 'custom.' (lower case) and may have additional
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/metric/IgniteMetrics.java:29]: Can we go with only one logic - either add `custom.` to all provided names or check if registryName starts with `custom.` and fail if not? I prefer first option - let's just add `custom.` to all provided `registryName`.

File: modules/core/src/main/java/org/apache/ignite/metric/MetricRegistry.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     * Registers an int metric which value will be queried from the specified supplier.
+     *
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param desc Metric description.
+     */
+    void register(String name, IntSupplier supplier, @Nullable String desc);
+
+    /**
+     * Registers a long which value will be queried from the specified supplier.
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/metric/MetricRegistry.java:47]: nit: Registers a long **metric** which value.

"metric" word lost

File: modules/core/src/main/java/org/apache/ignite/metric/MetricRegistry.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+     * @param name Metric name.
+     * @param supplier Metric value supplier.
+     * @param desc Metric description.
+     */
+    void register(String name, BooleanSupplier supplier, @Nullable String desc);
+
+    /**
+     * Removes metrics with the {@code name}.
+     *
+     * @param name Metric name..
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/metric/MetricRegistry.java:87]: nit: two dots in the en.

Can you, please, clarify - "name" expected to be full metric name like "custom.admin.sessions" or just "sessions".

**Comment by nizhikov** [comment]
@Vladsz83 Thanks for the updates - can you, please, clarify - you suggest to merge both PR's?

This and #11223 ?
Do we really need both?

**Comment by Vladsz83** [comment]
> @Vladsz83 Thanks for the updates - can you, please, clarify - you suggest to merge both PR's?
> 
> This and #11223 ? Do we really need both?

I vote for only #11223 . Minimal changes, doesn't break anything using `CdcConsumer` like https://github.com/apache/ignite-extensions/blob/master/modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/IgniteToKafkaCdcStreamer.java . @NSAmelchev , WDYT?

**Comment by nizhikov** [comment]
This PR includes #11223 and correct naming. 
Is this correct?

**Comment by Vladsz83** [comment]
> This PR includes #11223 and correct naming. Is this true?

Yes, but #11223 is significantly different. If merged, it doesn't require additional fixes of broken extensions and any user code which uses `CdcConsumer` and internal `class MetricRegistry`. I belive, we should remove usages of the internals from `CdcConsumer` first. Then, rename new `interface IgniteMetricRegistry` -> `interface MetricRegistry`. We could keep notes, TODOs about it in the code and the docs.

**Comment by nizhikov** [comment]
`CdcConsumer` must use public API only.
It's OK if extension will be fixed after this PR merged.

> Then, rename new interface IgniteMetricRegistry -> interface MetricRegistry

Let's not merge any temporarily public API in master.

**Comment by Vladsz83** [comment]
> `CdcConsumer` must use public API only. It's OK if extension will be fixed after this PR merged.

It uses `org.apache.ignite.internal.processors.metric.MetricRegistry` now.

> Let's not merge any temporarily public API in master.
Agree. But it is an experimental API.

**Comment by NSAmelchev** [comment]
@Vladsz83 Thank you for the fixes.

I suggest to proceed with this PR and close the #11223. So, the public CdcConsumer will use a public API. 

@nizhikov WDYT?

**Comment by nizhikov** [comment]
@Vladsz83 @NSAmelchev Will take a look shortly.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java [hunk 7]
```diff
@@ -745,4 +761,27 @@ private interface SunOperatingSystemMXBeanAccessor {
...
+            GridMetricManager.this.remove(customName(registryName), false);
+        }
+
+        /** {@inheritDoc} */
+        @NotNull @Override public Iterator<ReadOnlyMetricRegistry> iterator() {
+            return F.viewReadOnly(registries.values(), r -> r, r -> MetricUtils.customMetric(r.name())).iterator();
+        }
+
+        /** Adds {@link MetricUtils#CUSTOM_METRICS} to {@code name}. */
+        private String customName(String name) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/GridMetricManager.java:783]: Looks like `MetricUtils.customMetric` and `CustomMetricsImpl.customName` must be near each other.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/impl/MetricUtils.java [hunk 8]
```diff
@@ -45,22 +47,43 @@ public class MetricUtils {
...
     public static String metricName(String... names) {
-        assert names != null;
-        assert ensureAllNamesNotEmpty(names);
+        assert names != null && names.length > 0 : "Metric name must consist of at least one element.";
+
+        boolean custom = customMetric(names[0]);
 
-        if (names.length == 1)
-            return names[0];
+        for (int i = 0; i < names.length; i++) {
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/metric/impl/MetricUtils.java:70]: Do we really need that complex checks here? Can we move this check to `CustomMetricsImpl.customName`?

**Review by nizhikov — APPROVED** [review]

**Comment by nizhikov** [comment]
@Vladsz83 @NSAmelchev Leave few minor comments. Please, fix them prior to merge.

**Comment by nizhikov** [comment]
LGTM! Thank you.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/metric/impl/MetricUtils.java (hunk 8 above)
Comment (Vladsz83) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/metric/impl/MetricUtils.java:70]: We need some custom name checks in `MetricRegistryImpl#register(...)`. There is no call of `customName()`

**Pull request merged**